p2panda-net/src/engine/sync.rs

```diff
@@ -0,0 +1,277 @@
+            }
+            ToSyncActor::Accept { peer, connection } => {
+                self.on_accept_sync(peer, connection).await?
+            }
+            ToSyncActor::Shutdown => return Ok(false),
+        };
+
+        Ok(true)
+    }
+
+    /// Initiate a sync protocol session over a new bi-directional stream on the provided connections.
+    async fn on_open_sync(
+        &self,
+        peer: PublicKey,
+        topic: TopicId,
+        connection: Connection,
+    ) -> Result<()> {
+        debug!(
+            "Initiate sync session with peer {} over topic {:?}",
+            peer, topic
+        );
+
+        // Set up a channel for receiving new application messages.
+        let (tx, mut rx) = mpsc::channel(128);
+        let mut sink = PollSender::new(tx).sink_map_err(|e| SyncError::Protocol(e.to_string()));
+
+        // Spawn a task which opens a bi-directional stream over the provided connection and runs
+        // the sync protocol.
+        let protocol = self.sync_protocol.clone();
+        tokio::spawn(async move {
+            let result = async {
+                let (mut send, mut recv) = connection
+                    .open_bi()
+                    .await
+                    .map_err(|e| SyncError::Protocol(e.to_string()))?;
+
+                protocol
+                    .open(
+                        topic.as_bytes(),
+                        Box::new(&mut send),
+                        Box::new(&mut recv),
+                        Box::new(&mut sink),
+                    )
+                    .await
+            }
+            .await;
+
+            if let Err(err) = result {
+                error!("{err}");
+            };
+        });
+
+        // Spawn another task which picks up any new application messages and sends them
+        // on to the engine for handling.
+        let engine_actor_tx = self.engine_actor_tx.clone();
+        let mut sync_handshake_success = false;
+        tokio::spawn(async move {
+            while let Some(message) = rx.recv().await {
+                // We expect the first message to be a topic id
+                if let AppMessage::Topic(id) = &message {
+                    if sync_handshake_success {
+                        error!("topic already received from sync session");
+                        break;
+                    }
+                    sync_handshake_success = true;
+
+                    // Inform the engine that we are expecting sync messages from the peer on this topic
+                    engine_actor_tx
+                        .send(ToEngineActor::SyncHandshakeSuccess {
+                            peer,
+                            topic: id.to_owned().into(),
+                        })
+                        .await
+                        .expect("engine channel closed");
+
+                    continue;
+                }
+
+                let AppMessage::Bytes(bytes) = message else {
```

Comment:
Huh, I think it's my first time seeing syntax with this kind of `else` placement. Was a bit confused for a second.

Reply:
Yeh, it's pretty handy, and also funky at first :sweat_smile: